File: trappingRainWater/trappingRainWater.cpp
// clang++ trappingRainWater.cpp -std=c++11

#include <iostream>
#include <vector>
using namespace std;

class Solution{
public:
    int trap(vector<int>& height) {
        int res = 0;
        int size = height.size();
        if (size == 0)
            return 0;
        int left = 0;
        int right = size - 1;
        int leftMax = 0, maxRight = 0;
        while (left < right) {
            if (height[left] < height[right]) {
                if (height[left] > leftMax)
                    leftMax = height[left];
                else
                    res += leftMax - height[left];
                left++;
            } else {
                if (height[right] > maxRight)
                    maxRight = height[right];
                else
                    res += maxRight - height[right];
                right--;
            }
        }
        return res;
    }
};

int main()
{
    Solution s;
    vector<int> height = {0, 1, 0, 2, 1, 0, 1, 3, 2, 1, 2, 1};
    cout << s.trap(height) << endl;;
    return 0;
}

File: 42.trappingRainWater/trappingRainWater.java
class Solution {
    public int trap(int[] height) {
        int res = 0;
        int left = 0, right = height.length - 1;
        int leftM = 0, rightM = 0;
        while (left < right) {
            if (height[left] < height[right]) {
                if (height[left] > leftM) {
                    leftM = height[left];
                } else {
                    res += leftM - height[left];
                }
                left++;
            } else {
                if (height[right] > rightM) {
                    rightM = height[right];
                } else {
                    res += rightM - height[right];
                }
                right--;
            }
        }
        return res;
    }
}
